### PR TITLE
event cache: introduce the `Paginator` API

### PR DESCRIPTION
**Pull request by bnjbvr**
This introduces a new helper object to run arbitrary pagination requests, backwards- or forward-. At the moment they're disconnected from the event cache, although I've put the files there for future convenience, since at some point we'll want to merge the retrieved events with the cache (? maybe).

This little state machine makes it possible to retrieve the initial data, given an initial event id, using the /context endpoint; then allow stateful pagination using a paginator kind of API. Paginating in the timeline indicates whether we've reached the start/end of the timeline.

The test for the state subscription is quite extensive and makes sure the basic functionality works as intended.

Some testing helpers have been (re)introduced in the SDK crate, simplifying the code, and introducing a better `EventFactory` / `EventBuilder` pattern than the existing one in the `matrix-sdk-test` crate. In particular, this can make use of some types in `matrix-sdk`, notably `SyncTimelineEvent` and `TimelineEvent`, and I've found the API to be simpler to use as well.

Commit history is messy, so we might want to squash when merging.

Part of #3234.

### CODE REVIEW

File: crates/matrix-sdk/src/event_cache/paginator.rs [hunk 1]
```diff
@@ -0,0 +1,564 @@
...
+            prev_batch_token: Mutex::new(None),
+            next_batch_token: Mutex::new(None),
+        }
+    }
+
+    /// Check if the current state of the paginator matches the expected one.
+    fn check_state(&self, expected: PaginatorState) -> Result<(), PaginatorError> {
+        let actual = self.state.get();
+        if actual != expected {
+            return Err(PaginatorError::InvalidPreviousState { expected, actual });
```
Comment (andybalaam) [inline crates/matrix-sdk/src/event_cache/paginator.rs:149]: Personal preference, but I'd prefer an `else` on this `if` to avoid the `return`.

File: crates/matrix-sdk/src/event_cache/paginator.rs [hunk 2]
```diff
@@ -0,0 +1,564 @@
...
+    SdkError(#[source] crate::Error),
+}
+
+/// A stateful object to reach to an event, and then paginate backward and
+/// forward from it.
+///
+/// See also the module-level documentation.
+pub struct Paginator {
+    /// The room in which we're going to run the pagination.
+    room: Box<dyn PaginableRoom>,
```
Comment (andybalaam) [inline crates/matrix-sdk/src/event_cache/paginator.rs:74]: Curious why you went for `Box` instead of making this struct generic.

File: crates/matrix-sdk/src/event_cache/paginator.rs [hunk 3]
```diff
@@ -0,0 +1,564 @@
...
+    pub fn state(&self) -> Subscriber<PaginatorState> {
+        self.state.subscribe()
+    }
+
+    /// Starts the pagination from the initial event.
+    ///
+    /// Only works for fresh [`Paginator`] objects, which are in the
+    /// [`PaginatorState::Initial`] state.
+    pub async fn start_from(&self, event_id: &EventId) -> Result<StartFromResult, PaginatorError> {
+        self.check_state(PaginatorState::Initial)?;
```
Comment (andybalaam) [inline crates/matrix-sdk/src/event_cache/paginator.rs:164]: I don't suppose we can start when we create the struct, meaning we don't need `Initial` state?

File: crates/matrix-sdk/src/event_cache/paginator.rs [hunk 4]
```diff
@@ -0,0 +1,564 @@
...
+        // initial. This check makes sure there's at most one which can set the
+        // state to FetchingTargetEvent, preventing a race condition.
+        if self.state.set_if_not_eq(PaginatorState::FetchingTargetEvent).is_none() {
+            return Err(PaginatorError::InvalidPreviousState {
+                expected: PaginatorState::Initial,
+                actual: PaginatorState::FetchingTargetEvent,
+            });
+        }
+
+        // TODO: do we want to lazy load members?
```
Comment (andybalaam) [inline crates/matrix-sdk/src/event_cache/paginator.rs:176]: Members seem irrelevant here, so maybe not?

File: crates/matrix-sdk/src/event_cache/paginator.rs [hunk 5]
```diff
@@ -0,0 +1,564 @@
...
+        let lazy_load_members = true;
+
+        let response = self.room.event_with_context(event_id, lazy_load_members).await?;
+
+        let has_prev = response.prev_batch_token.is_some();
+        let has_next = response.next_batch_token.is_some();
```
Comment (andybalaam) [inline crates/matrix-sdk/src/event_cache/paginator.rs:182]: Maybe move these 2 `is_some` lines down to just before these variables are used?

Comment (andybalaam) [reply crates/matrix-sdk/src/event_cache/paginator.rs:182]: Oh, no, I see! You are about to change them.

Comment (andybalaam) [reply crates/matrix-sdk/src/event_cache/paginator.rs:182]: Can we name them somehow to make clear these are the old values?

File: crates/matrix-sdk/src/event_cache/paginator.rs [hunk 6]
```diff
@@ -0,0 +1,564 @@
...
+        // Consolidate the events into a linear timeline, topologically ordered.
+        // - the events before are returned in the reverse topological order: invert
+        //   them.
+        // - insert the target event, if set.
+        // - the events after are returned in the correct topological order.
+
+        let events = response
+            .events_before
+            .into_iter()
+            .rev()
```
Comment (andybalaam) [inline crates/matrix-sdk/src/event_cache/paginator.rs:197]: Hilarious that we need to do this :-)

File: crates/matrix-sdk/src/event_cache/paginator.rs [hunk 7]
```diff
@@ -0,0 +1,564 @@
...
+            .chain(response.event)
+            .chain(response.events_after)
+            .collect();
+
+        Ok(StartFromResult { events, has_prev, has_next })
+    }
+
+    /// Runs a backward pagination, from the current state of the object.
+    ///
+    /// Will return immediately if we did hit the start of the timeline before.
```
Comment (andybalaam) [inline crates/matrix-sdk/src/event_cache/paginator.rs:207]:
```suggestion
    /// Will return immediately if we have already hit the start of the timeline.
```

File: crates/matrix-sdk/src/event_cache/paginator.rs [hunk 8]
```diff
@@ -0,0 +1,564 @@
...
+    ///
+    /// May return an error if it's already paginating, or if the call to
+    /// /messages failed.
+    pub async fn paginate_backward(&self) -> Result<PaginationResult, PaginatorError> {
+        self.paginate(Direction::Backward).await
+    }
+
+    /// Runs a forward pagination, from the current state of the object.
+    ///
+    /// Will return immediately if we did hit the end of the timeline before.
```
Comment (andybalaam) [inline crates/matrix-sdk/src/event_cache/paginator.rs:217]:
```suggestion
    /// Will return immediately if we have already hit the end of the timeline.
```

File: crates/matrix-sdk/src/event_cache/paginator.rs [hunk 9]
```diff
@@ -0,0 +1,564 @@
...
+                if next_batch_token.is_none() {
+                    return Ok(PaginationResult { events: Vec::new(), hit_end_of_timeline: true });
+                };
+                next_batch_token.clone()
+            }
+        };
+
+        // Note: it's possible two callers have checked the state and both figured it's
+        // idle. This check makes sure there's at most one which can set the
+        // state to paginating, preventing a race condition.
```
Comment (andybalaam) [inline crates/matrix-sdk/src/event_cache/paginator.rs:248]: Are these kinds of errors expected to happen? If so, how will we handle them?

Comment (andybalaam) [reply crates/matrix-sdk/src/event_cache/paginator.rs:248]: Also, should the `check_state` call be as close as possible to `set_if_not_eq`?

File: crates/matrix-sdk/src/event_cache/paginator.rs [hunk 10]
```diff
@@ -0,0 +1,564 @@
...
+
+        let response_result =
+            self.room.messages(MessagesOptions::new(dir).from(token.as_deref())).await;
+
+        // In case of error, reset the state to idle.
+        let response = match response_result {
+            Ok(res) => res,
+            Err(err) => {
+                self.state.set(PaginatorState::Idle);
+                return Err(err);
```
Comment (andybalaam) [inline crates/matrix-sdk/src/event_cache/paginator.rs:264]: Is there some cool `thiserror` magic meaning we don't need `.into()` here?

File: crates/matrix-sdk/src/event_cache/paginator.rs [hunk 11]
```diff
@@ -0,0 +1,564 @@
...
+                hit
+            }
+            Direction::Forward => {
+                let hit = response.end.is_none();
+                *self.next_batch_token.lock().await = response.end;
+                hit
+            }
+        };
+
+        // TODO: what to do with state events?
```
Comment (andybalaam) [inline crates/matrix-sdk/src/event_cache/paginator.rs:281]: Some state events appear in the timeline (e.g. location shares) so maybe handle them the same as everything else?

File: crates/matrix-sdk/src/event_cache/paginator.rs [hunk 12]
```diff
@@ -0,0 +1,564 @@
...
+        // In case of error, reset the state to idle.
+        let response = match response_result {
+            Ok(res) => res,
+            Err(err) => {
+                self.state.set(PaginatorState::Idle);
+                return Err(err);
+            }
+        };
+
+        let hit_end_of_timeline = match dir {
```
Comment (andybalaam) [inline crates/matrix-sdk/src/event_cache/paginator.rs:268]: This structure of 2 methods for forward/backward, then one method for paginate, then `match`es inside that method, makes me feel a bit funny, but I can't see a good way out at the moment, unless you had an enum for the directions that had methods like `hit_end_of_timeline` on it? Not sure it would work, and would be a bit awkward for the early return on line 232.

Comment (andybalaam) [reply crates/matrix-sdk/src/event_cache/paginator.rs:268]: This method does feel slightly long to me, so if we could figure a way to break off into direction-specific submethods, that would be cool...

File: crates/matrix-sdk/src/event_cache/paginator.rs (hunk 10 above)
Comment (andybalaam) [reply crates/matrix-sdk/src/event_cache/paginator.rs:264]: Oh, no it's just wrapped on line 332.

File: crates/matrix-sdk/src/event_cache/paginator.rs [hunk 13]
```diff
@@ -0,0 +1,564 @@
...
+            _lazy_load_members: bool,
+        ) -> Result<EventWithContextResponse, PaginatorError> {
+            // Wait for the room to be marked as ready first.
+            self.room_ready.notified().await;
+
+            let event_factory = EventFactory::new().room(*ROOM_ID).sender(*USER_ID);
+
+            let event = event_factory.text_msg("hello!").event_id(event_id).into_raw_timeline();
+
+            let before = (0..10)
```
Comment (andybalaam) [inline crates/matrix-sdk/src/event_cache/paginator.rs:389]: Would it be worth numbering these backwards to help understand the order?

File: crates/matrix-sdk/src/event_cache/paginator.rs [hunk 14]
```diff
@@ -0,0 +1,564 @@
...
+                .rev()
+                .map(|i| {
+                    TimelineEvent::new(event_factory.text_msg(format!("{i}")).into_raw_timeline())
+                })
+                .collect();
+
+            let after = (10..20)
```
Comment (andybalaam) [inline crates/matrix-sdk/src/event_cache/paginator.rs:396]: Could start at 11 to indicate that `event` is number 10?

File: crates/matrix-sdk/src/event_cache/paginator.rs [hunk 15]
```diff
@@ -0,0 +1,564 @@
...
+            self.room_ready.notified().await;
+
+            let event_factory = EventFactory::new().room(*ROOM_ID).sender(*USER_ID);
+
+            let event = event_factory.text_msg("hello!").event_id(event_id).into_raw_timeline();
+
+            let before = (0..10)
+                .rev()
+                .map(|i| {
+                    TimelineEvent::new(event_factory.text_msg(format!("{i}")).into_raw_timeline())
```
Comment (andybalaam) [inline crates/matrix-sdk/src/event_cache/paginator.rs:392]: Personal taste, but I would name these e.g. "before-{i}" or similar and the same for after.

File: crates/matrix-sdk/src/event_cache/paginator.rs [hunk 16]
```diff
@@ -0,0 +1,564 @@
...
+        )
+        .await;
+
+        assert!(state.next().now_or_never().is_none());
+
+        // Mark the dummy room as ready. The query may now terminate.
+        room.mark_ready();
+
+        // After fetching the initial event data, the paginator switches to `Idle`.
+        assert_eq!(state.next().await, Some(PaginatorState::Idle));
```
Comment (andybalaam) [inline crates/matrix-sdk/src/event_cache/paginator.rs:506]: I would prefer a separate test for the content of the response. All this state stuff is very important but also very distracting from the "normal" case of "when I ask for event, I get events".

File: crates/matrix-sdk/src/event_cache/paginator.rs [hunk 17]
```diff
@@ -0,0 +1,564 @@
...
+                expected: real_expected,
+                actual: real_actual
+            }) = task.await
+        );
+        assert_eq!(real_expected, expected);
+        assert_eq!(real_actual, actual);
+    }
+
+    #[async_test]
+    async fn test_state() {
```
Comment (andybalaam) [inline crates/matrix-sdk/src/event_cache/paginator.rs:456]: Please can we have more tests for paginating multiple times (successfully), hitting the end, paginating forwards, and paginating in both directions?

**Review by andybalaam — CHANGES_REQUESTED** [review]
This looks great! Lots of questions and suggestions, and an ask for more tests. Thanks!

File: crates/matrix-sdk/src/event_cache/paginator.rs (hunk 2 above)
Comment (bnjbvr) [reply crates/matrix-sdk/src/event_cache/paginator.rs:74]: Force of habit, and it's not performance critical, so the indirect call seemed fine.

File: crates/matrix-sdk/src/event_cache/paginator.rs (hunk 3 above)
Comment (bnjbvr) [reply crates/matrix-sdk/src/event_cache/paginator.rs:164]: I wanted to distinguish "creating a data structure" = filling the fields, something that's infallible, from "starting the state machine" = fallible attempt to get some data from network or disk.

File: crates/matrix-sdk/src/event_cache/paginator.rs (hunk 4 above)
Comment (bnjbvr) [reply crates/matrix-sdk/src/event_cache/paginator.rs:176]: They are relevant! One may click a permalink to a room they've never visited, and as a result have no information whatsoever with respect to the member profiles, so we'd need to process the room member data there. Leaving for later.

File: crates/matrix-sdk/src/event_cache/paginator.rs (hunk 9 above)
Comment (bnjbvr) [reply crates/matrix-sdk/src/event_cache/paginator.rs:248]: > Are these kinds of errors expected to happen? If so, how will we handle them?

I don't understand these two questions. There are no errors that can happen, for the reasons explained in the code comment.

> Also, should the check_state call be as close as possible to set_if_not_eq?

No, because it'd let one believe that `paginate_backward()` works and returns nothing before calling `start_from()` (since the `prev_batch_token` would be none, and we'd short-circuit the return).

File: crates/matrix-sdk/src/event_cache/paginator.rs (hunk 11 above)
Comment (bnjbvr) [reply crates/matrix-sdk/src/event_cache/paginator.rs:281]: (Indeed, and related to the comment about lazy member loading above — deferred to later)

File: crates/matrix-sdk/src/event_cache/paginator.rs (hunk 12 above)
Comment (bnjbvr) [reply crates/matrix-sdk/src/event_cache/paginator.rs:268]: How would you split that up? This method is 60 lines of code long (including 14 empty lines or comment lines), and the duplication in the matches seems less terrible than duplicating the entire methods, so I'm all up for ideas!

File: crates/matrix-sdk/src/event_cache/paginator.rs (hunk 3 above)
Comment (andybalaam) [reply crates/matrix-sdk/src/event_cache/paginator.rs:165]: Does that distinction buy us something? E.g. might we retry `start_from` later? If so, it's worth it.

File: crates/matrix-sdk/src/event_cache/paginator.rs (hunk 5 above)
Comment (andybalaam) [reply crates/matrix-sdk/src/event_cache/paginator.rs:183]: Ah, I was confused.

Then I go back to my original comment: why not create these variables immediately before they are used instead of creating them a few lines earlier? (Very minor point)

File: crates/matrix-sdk/src/event_cache/paginator.rs (hunk 9 above)
Comment (andybalaam) [reply crates/matrix-sdk/src/event_cache/paginator.rs:250]: > No, because it'd let one believe that paginate_backward() works and returns nothing before calling start_from() (since the prev_batch_token would be none, and we'd short-circuit the return).

I guess we could get the state to `Paginating` early, and unset it during the early returns, but it's probably not worth it. I'm still wondering whether we think this race where we check the state and then it changes while we are running this code is going to happen. If so, the early return might be misleading, right?

File: crates/matrix-sdk/src/event_cache/paginator.rs (hunk 12 above)
Comment (andybalaam) [reply crates/matrix-sdk/src/event_cache/paginator.rs:270]: OK, so I tried it out just to see, and it's not too bad: https://github.com/matrix-org/matrix-rust-sdk/pull/3325/files

Comment (andybalaam) [reply crates/matrix-sdk/src/event_cache/paginator.rs:270]: If `DirThing` had a good name to express what it was, maybe this would be ... good?

Very happy to leave it as it is, just thinking out loud.

Comment (bnjbvr) [reply crates/matrix-sdk/src/event_cache/paginator.rs:270]: I don't think introducing 30 new lines and a new data structure instead of keeping the 66-lines long method as is is a net positive, so I'll keep the method as it is right now. Or I can just pass a `&RwLock<Option<String>>` to this `paginate` helper :-)

File: crates/matrix-sdk/src/event_cache/paginator.rs (hunk 9 above)
Comment (bnjbvr) [reply crates/matrix-sdk/src/event_cache/paginator.rs:250]: > I mean the Err(PaginatorError::InvalidPreviousState that we return. Is this expected to happen?

Spawn two tasks, each calling the method, while the initial state is `Idle`? It's a race condition, and it would likely require total control over tokio to trigger…

> If so, the early return might be misleading, right?

I don't think it is: having the initial state be "idle" is an invariant, so should be checked at all times; then having two callers race on it is rather exceptional, and what the second check is preventing.

File: crates/matrix-sdk/src/event_cache/paginator.rs (hunk 5 above)
Comment (bnjbvr) [reply crates/matrix-sdk/src/event_cache/paginator.rs:183]: Because that would require taking the lock again, since the `response`'s ownership is partially taken when assigning the inner fields on the next two lines.

File: crates/matrix-sdk/src/event_cache/paginator.rs (hunk 3 above)
Comment (bnjbvr) [reply crates/matrix-sdk/src/event_cache/paginator.rs:165]: Yes.

File: crates/matrix-sdk/src/event_cache/paginator.rs (hunk 12 above)
Comment (andybalaam) [reply crates/matrix-sdk/src/event_cache/paginator.rs:270]: Agree, was fun trying it out :-)

File: crates/matrix-sdk/src/event_cache/paginator.rs (hunk 13 above)
Comment (bnjbvr) [reply crates/matrix-sdk/src/event_cache/paginator.rs:391]: I've checked, and turns out `9..=0` is an empty range — what the duck, Rust?

**Comment by bnjbvr** [comment]
I've now added isolated tests for getting events from start_from / paginate_backwards / paginate_forwards, and I think it covers all the things you've asked — PTAL!

File: crates/matrix-sdk/src/event_cache/paginator.rs [hunk 18]
```diff
@@ -221,28 +221,22 @@ impl Paginator {
...
     pub async fn paginate_forward(&self) -> Result<PaginationResult, PaginatorError> {
-        self.paginate(Direction::Forward).await
+        self.paginate(Direction::Forward, &self.next_batch_token).await
     }
 
-    async fn paginate(&self, dir: Direction) -> Result<PaginationResult, PaginatorError> {
+    async fn paginate(
+        &self,
+        dir: Direction,
+        token_lock: &Mutex<Option<String>>,
```
Comment (andybalaam) [inline crates/matrix-sdk/src/event_cache/paginator.rs:230]: This is great!

**Review by andybalaam — APPROVED** [review]
Brilliant, thank you!
